File: skillplate/webhooks.py
"""Verifica delle firme dei webhook Skillplate.

I payload sono firmati HMAC-SHA256 e arrivano nell'header
``X-Skillplate-Signature``. Verifica sempre prima di fidarti del contenuto.
"""

import hashlib
import hmac
import json

SIGNATURE_HEADER = "X-Skillplate-Signature"

EVENTI = (
    "payment.succeeded",
    "payment.failed",
    "subscription.started",
    "subscription.cancelled",
    "user.created",
    "user.updated",
    "lesson.completed",
    "module.completed",
    "course.completed",
)


def compute_signature(payload, secret):
    """HMAC-SHA256 esadecimale del corpo grezzo della richiesta."""
    if isinstance(payload, str):
        payload = payload.encode("utf-8")
    if isinstance(secret, str):
        secret = secret.encode("utf-8")
    return hmac.new(secret, payload, hashlib.sha256).hexdigest()


def verify_signature(payload, signature, secret):
    """Confronto a tempo costante fra firma attesa e header ricevuto.

    Accetta sia ``<hex>`` sia il formato prefissato ``sha256=<hex>``.
    """
    if not signature:
        return False
    ricevuta = signature.strip()
    if "=" in ricevuta:
        prefisso, _, resto = ricevuta.partition("=")
        if prefisso.lower() == "sha256":
            ricevuta = resto.strip()
    return hmac.compare_digest(compute_signature(payload, secret), ricevuta)


def parse_event(payload, signature, secret):
    """Verifica la firma e restituisce l'evento decodificato.

    Solleva ``ValueError`` se la firma non torna: il corpo non va usato.
    """
    if not verify_signature(payload, signature, secret):
        raise ValueError("Firma webhook non valida: payload scartato")
    if isinstance(payload, bytes):
        payload = payload.decode("utf-8")
    return json.loads(payload)


def product_ids(evento):
    """Estrae i product_id da un payload ``payment.succeeded`` / ``payment.failed``."""
    voci = evento.get("items") or evento.get("data", {}).get("items") or []
    return [voce.get("product_id") for voce in voci if voce.get("product_id")]

File: node/src/webhooks.js
/** Verifica delle firme dei webhook Skillplate.
 *
 * I payload sono firmati HMAC-SHA256 e arrivano nell'header
 * `X-Skillplate-Signature`. Verifica sempre prima di fidarti del contenuto.
 */

import crypto from "node:crypto";

export const SIGNATURE_HEADER = "X-Skillplate-Signature";

export const EVENTI = [
  "payment.succeeded",
  "payment.failed",
  "subscription.started",
  "subscription.cancelled",
  "user.created",
  "user.updated",
  "lesson.completed",
  "module.completed",
  "course.completed",
];

/** HMAC-SHA256 esadecimale del corpo grezzo della richiesta. */
export function computeSignature(payload, secret) {
  return crypto.createHmac("sha256", secret).update(payload).digest("hex");
}

/** Confronto a tempo costante fra firma attesa e header ricevuto.
 *
 * Accetta sia `<hex>` sia il formato prefissato `sha256=<hex>`.
 */
export function verifySignature(payload, signature, secret) {
  if (!signature) return false;
  let ricevuta = String(signature).trim();
  const separatore = ricevuta.indexOf("=");
  if (separatore !== -1 && ricevuta.slice(0, separatore).toLowerCase() === "sha256") {
    ricevuta = ricevuta.slice(separatore + 1).trim();
  }
  const attesa = computeSignature(payload, secret);
  // timingSafeEqual richiede buffer di pari lunghezza
  if (ricevuta.length !== attesa.length) return false;
  return crypto.timingSafeEqual(Buffer.from(attesa, "utf8"), Buffer.from(ricevuta, "utf8"));
}

/** Verifica la firma e restituisce l'evento decodificato.
 *
 * Solleva `Error` se la firma non torna: il corpo non va usato.
 */
export function parseEvent(payload, signature, secret) {
  if (!verifySignature(payload, signature, secret)) {
    throw new Error("Firma webhook non valida: payload scartato");
  }
  return JSON.parse(Buffer.isBuffer(payload) ? payload.toString("utf8") : payload);
}

/** Estrae i product_id da un payload `payment.succeeded` / `payment.failed`. */
export function productIds(evento) {
  const voci = evento.items ?? evento.data?.items ?? [];
  return voci.map((voce) => voce.product_id).filter(Boolean);
}
